runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java

```diff
@@ -443,7 +443,7 @@ public boolean optimisticallyOwnedByThreadUnsafe() {
      * @param optimisticStream The new optimistic stream to install.
      */
     public void setUncommittedChanges(WriteSetSMRStream optimisticStream) {
-        if (this.optimisticStream != null) {
+        if (this.optimisticStream != null && !isMonotonicObject()) {
             rollbackUncommittedChangesUnsafe();
         }
         this.optimisticStream = optimisticStream;
```

Comment:
It looks like the only reason we update the `optimisticStream` here is to eventually sync through it and set the upcall results to null.  However, it seems like this computation will be redone for every context switch. Can the `isMonotonicObject()` check can be moved to the caller `setUncommittedChanges()`? Then upcall related results can be directly set on log updates.

Reply:
Thanks, fixed.